--- a/systems/bde/linux/kernel/linux-kernel-bde.c
+++ b/systems/bde/linux/kernel/linux-kernel-bde.c
@@ -2517,6 +2517,16 @@ _pci_probe(struct pci_dev *dev, const st
             }
         }
 
+        if ((PCI_FIND_DEV(BROADCOM_VENDOR_ID, BCM56160_DEVICE_ID, NULL) != NULL) ||
+            (PCI_FIND_DEV(BROADCOM_VENDOR_ID, BCM56172_DEVICE_ID, NULL) != NULL)) {
+            if (maxpayload) {
+                maxpayload = 128;
+                if (debug >= 1) gprintk("force max payload size to 128\n");
+            }
+            use_msi = 1;
+            if (debug >= 1) gprintk("force usemsi to 1\n");
+        }
+
         if (forceirq > 0 && dev->irq != (uint32) forceirq) {
             if (forceirqubm & (1U << (_ndevices - 1))) {
                 dev->irq = forceirq;
