--- a/systems/bde/linux/kernel/linux_dma.c
+++ b/systems/bde/linux/kernel/linux_dma.c
@@ -962,12 +962,8 @@ _sinval(int d, void *ptr, int length)
 #if defined(dma_cache_wback_inv)
      dma_cache_wback_inv((unsigned long)ptr, length);
 #else
-#if defined(IPROC_CMICD) || defined(BCM958525)
     /* FIXME: need proper function to replace dma_cache_sync */
     dma_sync_single_for_cpu(NULL, (unsigned long)ptr, length, DMA_BIDIRECTIONAL);
-#else
-    dma_cache_sync(NULL, ptr, length, DMA_BIDIRECTIONAL);
-#endif
 #endif
     return 0;
 }
@@ -978,12 +974,8 @@ _sflush(int d, void *ptr, int length)
 #if defined(dma_cache_wback_inv)
     dma_cache_wback_inv((unsigned long)ptr, length);
 #else
-#if defined(IPROC_CMICD) || defined(BCM958525)
     /* FIXME: need proper function to replace dma_cache_sync */
     dma_sync_single_for_cpu(NULL, (unsigned long)ptr, length, DMA_BIDIRECTIONAL);
-#else
-    dma_cache_sync(NULL, ptr, length, DMA_BIDIRECTIONAL);
-#endif
 #endif
 
     return 0;
