--- a/systems/linux/kernel/modules/shared/gmodule.c
+++ b/systems/linux/kernel/modules/shared/gmodule.c
@@ -147,6 +147,15 @@ static int _gmodule_proc_release(struct
     return single_release(inode, file);
 }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(5,6,0)
+struct proc_ops _gmodule_proc_fops = {
+    proc_open:       _gmodule_proc_open,
+    proc_read:       seq_read,
+    proc_lseek:      seq_lseek,
+    proc_write:      _gmodule_proc_write,
+    proc_release:    _gmodule_proc_release,
+};
+#else
 struct file_operations _gmodule_proc_fops = {
     owner:      THIS_MODULE,
     open:       _gmodule_proc_open,
@@ -155,6 +164,7 @@ struct file_operations _gmodule_proc_fop
     write:      _gmodule_proc_write,
     release:    _gmodule_proc_release,
 };
+#endif
 #else
 int
 gmodule_vpprintf(char** page_ptr, const char* fmt, va_list args)
